server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -334,32 +334,88 @@ protected FieldMapper clone() {
+        }
+        if (fieldType.omitNorms() && !other.omitNorms()) {
+            conflicts.add("mapper [" + name() + "] has different [norms] values, cannot change from disable to enabled");
+        }
+        if (fieldType.storeTermVectors() != other.storeTermVectors()) {
+            conflicts.add("mapper [" + name() + "] has different [store_term_vector] values");
+        }
+        if (fieldType.storeTermVectorOffsets() != other.storeTermVectorOffsets()) {
+            conflicts.add("mapper [" + name() + "] has different [store_term_vector_offsets] values");
+        }
+        if (fieldType.storeTermVectorPositions() != other.storeTermVectorPositions()) {
+            conflicts.add("mapper [" + name() + "] has different [store_term_vector_positions] values");
+        }
+        if (fieldType.storeTermVectorPayloads() != other.storeTermVectorPayloads()) {
+            conflicts.add("mapper [" + name() + "] has different [store_term_vector_payloads] values");
+        }
+
+        // null and "default"-named index analyzers both mean the default is used
+        if (fieldType.indexAnalyzer() == null || "default".equals(fieldType.indexAnalyzer().name())) {
+            if (other.indexAnalyzer() != null && "default".equals(other.indexAnalyzer().name()) == false) {
+                conflicts.add("mapper [" + name() + "] has different [analyzer]");
             }
-            throw new IllegalArgumentException("mapper [" + fieldType().name() + "] of different type, current_type [" + contentType()
-                + "], merged_type [" + mergedType + "]");
+        } else if (other.indexAnalyzer() == null || "default".equals(other.indexAnalyzer().name())) {
+            conflicts.add("mapper [" + name() + "] has different [analyzer]");
+        } else if (fieldType.indexAnalyzer().name().equals(other.indexAnalyzer().name()) == false) {
+            conflicts.add("mapper [" + name() + "] has different [analyzer]");
+        }
+
+        if (Objects.equals(fieldType.similarity(), other.similarity()) == false) {
+            conflicts.add("mapper [" + name() + "] has different [similarity]");
         }
-        FieldMapper fieldMergeWith = (FieldMapper) mergeWith;
-        multiFields = multiFields.merge(fieldMergeWith.multiFields);
 
-        // apply changeable values
-        this.fieldType = fieldMergeWith.fieldType;
-        this.copyTo = fieldMergeWith.copyTo;
+        mergeOptions(mergeWith, conflicts);
+
+        if (conflicts.isEmpty()) {
+            multiFields = multiFields.merge(mergeWith.multiFields);
+            // apply changeable values
+            this.fieldType = mergeWith.fieldType;
+            this.copyTo = mergeWith.copyTo;
+        }
     }
 
+    /**
+     * Merge type-specific options and check for incompatible settings in mappings to be merged
+     */
+    protected abstract void mergeOptions(FieldMapper other, List<String> conflicts);
```

Comment:
Great that this is abstract, I like how it encourages implementors to think through what can be updated vs. not.